### PR TITLE
[SPARK-38521][SQL] Change `partitionOverwriteMode` from string to variable in Scala

### PR DESCRIPTION
**Pull request by jackylee-ch**
### Why are the changes needed?
We use `partitionOverwriteMode` in several places, but there is no unified variable definition. This PR is to change all this related modifications to a unified `DataSourceUtils. PARTITION_OVERWRITE_MODE` definition.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Origin UT

### CODE REVIEW

**Comment by jackylee-ch** [comment]
cc @cloud-fan 
BTW, it seems we support partition overwrite with `static mode` by default while partition manged by catalog. It looks weird to me as `staticmode` behaves catastrophically and deletes all data in table, maybe setting partition overwrite with `dynamic mode` by default while partition manged by catalog?
If ok, I will open another PR to fix this.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceUtils.scala [hunk 1]
```diff
@@ -46,6 +46,7 @@ object DataSourceUtils extends PredicateHelper {
    * The key to use for storing partitionBy columns as options.
    */
   val PARTITIONING_COLUMNS_KEY = "__partition_columns"
+  val PARTITION_OVERWRITE_MODE = "partitionOverwriteMode"
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceUtils.scala:49]: Can we add short doc for this one?

Comment (jackylee-ch) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceUtils.scala:49]: Sure, I will do that

**Review by HyukjinKwon — APPROVED** [review]

**Review by cloud-fan — APPROVED** [review]

**Review by wangyum — APPROVED** [review]

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceUtils.scala [hunk 2]
```diff
@@ -46,6 +46,11 @@ object DataSourceUtils extends PredicateHelper {
    * The key to use for storing partitionBy columns as options.
    */
   val PARTITIONING_COLUMNS_KEY = "__partition_columns"
+  /**
+   * The key to use for specifying partition overwrite mode when
+   * INSERT OVERWRITE a partitioned data source table.
+   */
+  val PARTITION_OVERWRITE_MODE = "partitionOverwriteMode"
```
Comment (huaxingao) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceUtils.scala:53]: What is the difference between this `DataSourceUtils.PARTITION_OVERWRITE_MODE` and `SQLConf.PARTITION_OVERWRITE_MODE`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceUtils.scala [hunk 3]
```diff
@@ -46,6 +46,11 @@ object DataSourceUtils extends PredicateHelper {
    * The key to use for storing partitionBy columns as options.
    */
   val PARTITIONING_COLUMNS_KEY = "__partition_columns"
+  /**
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceUtils.scala:49]: Could you add an empty line before this, @jackylee-ch ?

**Review by dongjoon-hyun — APPROVED** [review]

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceUtils.scala (hunk 2 above)
Comment (jackylee-ch) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceUtils.scala:53]: Basiclly, we will use `DataSourceUtils.PARTITION_OVERWRITE_MODE` if user set `partitionOverwriteMode` in table properties or options. If not, then we will read `SQLConf.PARTITION_OVERWRITE_MODE`.

**Comment by dongjoon-hyun** [comment]
Merged to master. Thank you, @jackylee-ch and all!

**Pull request closed without merging**